misc/memory-connection-limits/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+/// # use libp2p_memory_connection_limits as memory_connection_limits;
+///
+/// #[derive(NetworkBehaviour)]
+/// # #[behaviour(prelude = "libp2p_swarm::derive_prelude")]
+/// struct MyBehaviour {
+///   identify: identify::Behaviour,
+///   limits: memory_connection_limits::Behaviour
+/// }
+/// ```
+pub struct Behaviour {
+    max_process_memory_usage_bytes: Option<usize>,
+    max_process_memory_usage_percentage: Option<f64>,
+    system_physical_memory_bytes: usize,
+}
+
+impl Behaviour {
+    /// Sets the process memory usage threshold in absolute bytes.
+    ///
+    /// New inbound and outbound connections will be denied when the threshold is reached.
+    pub fn with_max_bytes(bytes: usize) -> Self {
+        let mut b = Self::new();
+        b.update_max_bytes(bytes);
+        b
+    }
+
+    /// Sets the process memory usage threshold in the percentage of the total physical memory,
+    /// all pending connections will be dropped when the threshold is exeeded
+    pub fn with_max_percentage(percentage: f64) -> Self {
+        let mut b = Self::new();
+        b.update_max_percentage(percentage);
+        b
+    }
+
+    /// Updates the process memory usage threshold in bytes,
+    pub fn update_max_bytes(&mut self, bytes: usize) {
+        self.max_process_memory_usage_bytes = Some(bytes);
+    }
+
+    /// Updates the process memory usage threshold in the percentage of the total physical memory,
+    pub fn update_max_percentage(&mut self, percentage: f64) {
+        self.max_process_memory_usage_percentage = Some(percentage);
+    }
+
+    fn new() -> Self {
+        use sysinfo::{RefreshKind, SystemExt};
+
+        let system_info = sysinfo::System::new_with_specifics(RefreshKind::new().with_memory());
+
+        Self {
+            max_process_memory_usage_bytes: None,
+            max_process_memory_usage_percentage: None,
+            system_physical_memory_bytes: system_info.total_memory() as usize,
+        }
+    }
+
+    fn check_limit(&self) -> Result<(), ConnectionDenied> {
+        if let Some(max_allowed_bytes) = self.max_allowed_bytes() {
+            if let Some(stats) = memory_stats::memory_stats() {
+                if stats.physical_mem > max_allowed_bytes {
+                    return Err(ConnectionDenied::new(MemoryUsageLimitExceeded {
+                        process_physical_memory_bytes: stats.physical_mem,
+                        max_allowed_bytes,
+                    }));
+                }
+            } else {
+                log::warn!("Failed to retrive process memory stats");
+            }
+        }
+
+        Ok(())
+    }
+
+    fn max_allowed_bytes(&self) -> Option<usize> {
+        let max_process_memory_usage_percentage = self
+            .max_process_memory_usage_percentage
+            .map(|p| (self.system_physical_memory_bytes as f64 * p).round() as usize);
+        match (
+            self.max_process_memory_usage_bytes,
+            max_process_memory_usage_percentage,
+        ) {
```

Comment:
Would it make sense to already compute `max_allowed_bytes` when the user sets the value in the constructor? i.e. only have one field in the behaviour instead of two `Option`s.

I think that would make this easier to reason about because we don't have to deal with `Option`s at all.

Reply:
Fixed